Introduce and start using the MallocSizeOf trait.

nnethercote opened the pull request:
MallocSizeOf is similar to the existing HeapSizeOf trait from the
heapsize crate. The only difference is that MallocSizeOf's
malloc_size_of_children() function takes an additional MallocSizeOfFn
argument, which is used to measure heap blocks. This extra argument
makes MallocSizeOf match how Gecko's memory measurements work, and is
required for Stylo to integrate with DMD.

The patch also introduces a second trait, MallocSizeOfWithGuard, which
is much the same as MallocSizeOf, but with a |guard| argument for the
global style lock.

Finally, the patch uses the new traits to measure a small amount of
Stylo's memory usage.




---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes OR
- [X] These changes do not require tests because code is only for Gecko integration.

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/17044)
<!-- Reviewable:end -->

nnethercote committed:
Introduce and start using the MallocSizeOf trait.

MallocSizeOf is similar to the existing HeapSizeOf trait from the
heapsize crate. The only difference is that MallocSizeOf's
malloc_size_of_children() function takes an additional MallocSizeOfFn
argument, which is used to measure heap blocks. This extra argument
makes MallocSizeOf match how Gecko's memory measurements work, and is
required for Stylo to integrate with DMD.

The patch also introduces a second trait, MallocSizeOfWithGuard, which
is much the same as MallocSizeOf, but with a |guard| argument for the
global style lock.

Finally, the patch uses the new traits to measure a small amount of
Stylo's memory usage.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/gecko/generated/bindings.rs, components/style/stylesheets.rs, components/style/properties/declaration_block.rs, components/style/properties/properties.mako.rs
 * @emilio: components/style/gecko/generated/bindings.rs, components/style/stylesheets.rs, components/style/properties/declaration_block.rs, components/style/properties/properties.mako.rs, ports/geckolib/glue.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify style code, but no tests are modified. Please consider adding a test!

nnethercote commented:
@heycam: This is the Servo part of the code you reviewed in https://bugzilla.mozilla.org/show_bug.cgi?id=1353998. The only changes are:
- I addressed your comments.
- I added handling of empty allocations (with pointer value of 0x1) as @emilio suggested.
- I added `#[cfg(feature = "gecko")]` throughout.

jdm commented:
0x1 no longer works; we need to use the same technique as https://github.com/servo/heapsize/pull/81.

nnethercote commented:
Thanks, @jdm. I see that Servo is still using heapsize 0.3.9, and so hasn't received that change.

jdm commented:
```
error[E0425]: cannot find function `align_of` in this scope

   --> /home/travis/build/servo/servo/components/style/stylesheets.rs:112:24

    |

112 |     if ptr as usize <= align_of::<T>() {

    |                        ^^^^^^^^^^^^^ not found in this scope

    |

    = help: possible candidate is found in another module, you can import it into scope:

              `use bitflags::__core::mem::align_of;`

error[E0308]: mismatched types

   --> /home/travis/build/servo/servo/components/style/stylesheets.rs:115:24

    |

115 |         malloc_size_of(ptr)

    |                        ^^^ expected enum `std::os::raw::c_void`, found type parameter

    |

    = note: expected type `*const std::os::raw::c_void`

               found type `*const T`

    = help: here are some functions which might fulfill your needs:

            - .offset(...)

            - .wrapping_offset(...)

error: aborting due to previous error

error: Could not compile `style`.


```

emilio commented on components/style/stylesheets.rs:
```diff
@@ -99,6 +103,64 @@ pub struct Namespaces {
+pub type MallocSizeOfFn = unsafe extern "C" fn(ptr: *const c_void) -> usize;
+
+/// Call malloc_size_of on ptr, first checking that the allocation isn't empty.
+#[cfg(feature = "gecko")]
```
Is there a reason this all needs to be `cfg`'d out? If we can compile it in servo mode too it'd be better IMO (less differences among builds).

nnethercote replied:
Just that it's not used in Servo.

emilio replied:
I'd prefer to not compile it out if possible, even if it's not used, because if someone is changing related code and causes this not to compile would only see the error when compiling the geckolib version.

emilio commented:
That being said, this was reviewed already no? If you could do that bit, that'd be great. Otherwise, go ahead and land it as you wish :)

@bors-servo delegate+

bors-servo commented:
:v: @nnethercote can now approve this pull request

nnethercote commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 6d5b124 has been approved by `nnethercote`

<!-- @bors-servo r=nnethercote 6d5b1242dbcaf00415a925de14c8f3c31163d920 -->

bors-servo commented:
:hourglass: Testing commit 6d5b1242dbcaf00415a925de14c8f3c31163d920 with merge c558c55058721a856b29c3a79cea49e53bbc4a84...

bors-servo commented:
:bulb: This pull request was already approved, no need to approve it again.

- This pull request is currently being tested. If there's no response from the continuous integration service, you may use `retry` to trigger a build again.

bors-servo commented:
:pushpin: Commit 6d5b124 has been approved by `nnethercote`

<!-- @bors-servo r=nnethercote 6d5b1242dbcaf00415a925de14c8f3c31163d920 -->

bors-servo commented:
:hourglass: Testing commit 6d5b1242dbcaf00415a925de14c8f3c31163d920 with merge 1ec79b6756e012ad2c59264912b2774f103f3aef...

nnethercote commented:
@bors-servo: r=emilio

bors-servo commented:
:bulb: This pull request was already approved, no need to approve it again.

- This pull request is currently being tested. If there's no response from the continuous integration service, you may use `retry` to trigger a build again.

bors-servo commented:
:pushpin: Commit 6d5b124 has been approved by `emilio`

<!-- @bors-servo r=emilio 6d5b1242dbcaf00415a925de14c8f3c31163d920 -->

bors-servo commented:
:hourglass: Testing commit 6d5b1242dbcaf00415a925de14c8f3c31163d920 with merge aca09436b71e6defe007e1d17a9d79214d094a7b...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/7197), [arm32](http://build.servo.org/builders/arm32/builds/7412), [arm64](http://build.servo.org/builders/arm64/builds/7449), [linux-dev](http://build.servo.org/builders/linux-dev/builds/7542), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/4297), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/4316), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/7520), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/5), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/5), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/4226), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/4291), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/5), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/6), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/2764)
Approved by: emilio
Pushing aca09436b71e6defe007e1d17a9d79214d094a7b to master...
